src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,403 @@
+	midiclock.signal_handler().set(FUNC(yamaha_dx9_state::midiclock_w));
+
+	MIDI_PORT(config, "mdin", midiin_slot, "midiin").rxd_handler().set(FUNC(yamaha_dx9_state::midi_r));
+
+	auto &mdout(MIDI_PORT(config, "mdout", midiout_slot, "midiout"));
+	m_maincpu->out_ser_tx_cb().set(mdout, FUNC(midi_port_device::write_txd));
+
+	// Configure the LCD screen.
+	screen_device &lcd_screen(SCREEN(config, "lcd_screen", SCREEN_TYPE_LCD));
+	lcd_screen.set_refresh_hz(60);
+	lcd_screen.set_vblank_time(ATTOSECONDS_IN_USEC(2500)); /* not accurate */
+	lcd_screen.set_screen_update("lcdc", FUNC(hd44780_device::screen_update));
+	lcd_screen.set_size(6*16+1, 10*2+1);
+	lcd_screen.set_visarea_full();
+	lcd_screen.set_palette("palette");
+
+	PALETTE(config, "palette", FUNC(yamaha_dx9_state::palette_init), 3);
+
+	hd44780_device &lcdc(HD44780(config, "lcdc", 0));
+	lcdc.set_lcd_size(2, 16);
+	lcdc.set_pixel_update_cb(FUNC(yamaha_dx9_state::lcd_pixel_update));
+
+	config.set_default_layout(layout_dx9);
+}
+
+
+/**
+ * yamaha_dx9_state::ops_w
+ */
+void yamaha_dx9_state::ops_w(offs_t offset, uint8_t data)
+{
+	// printf("OPS: %02X=%02X\n", offset, data);
+}
+
+
+/**
+ * yamaha_dx9_state::egs_w
+ */
+void yamaha_dx9_state::egs_w(offs_t offset, uint8_t data)
+{
+	// printf("EGS: %02X=%02X\n", offset, data);
+}
+
+
+/**
+ * yamaha_dx9_state::led_w
+ */
+void yamaha_dx9_state::led_w(offs_t offset, uint8_t data)
+{
+	// Since the memory mapped region that calls this function is only two byts in
+	// size, the led number is the least-significant bit of the offset.
+	// The DX9's LEDs are wired so that a high input line disables the segment, so
+	// get the one's complement of the data.
+	m_leds[offset & 1] = (~data) & 0xFF;
+}
+
+
+/**
+ * yamaha_dx9_state::p1_r
+ */
+uint8_t yamaha_dx9_state::p1_r(offs_t offset)
+{
+	// The ADC EOC line is wired to bit 4, as well as the Cassette Interface input, which
+	// is wired to bit 7. This is currently not fully implemented.
+	return m_adc->eoc_r() << 4;
+}
+
+
+/**
+ * yamaha_dx9_state::p1_w
+ */
+void yamaha_dx9_state::p1_w(offs_t offset, uint8_t data)
+{
+	// The low-nibble is written by the firmware to select the key/switch driver input line.
+	m_key_switch_input_select = data & 0xF;
+
+	// The cassette interface remote port polarity is set by bit 5.
+	m_cassette_interface_remote_polarity = data & 0x20;
+
+	// The cassette interface output polarity is set by bit 6.
+	m_cassette_interface_output_polarity = data & 0x40;
+}
+
+
+static INPUT_PORTS_START(dx9)
+	PORT_START("KEY_SWITCH_INPUT.0")
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_UNKNOWN) PORT_NAME("Yes/Up")
+	PORT_BIT(0x02, IP_ACTIVE_HIGH, IPT_UNKNOWN) PORT_NAME("No/Down")
+	PORT_BIT(0x04, IP_ACTIVE_HIGH, IPT_UNKNOWN) PORT_NAME("Store")
+	PORT_BIT(0x10, IP_ACTIVE_HIGH, IPT_UNKNOWN) PORT_NAME("Function")
+	PORT_BIT(0x20, IP_ACTIVE_HIGH, IPT_UNKNOWN) PORT_NAME("Edit")
+	PORT_BIT(0x40, IP_ACTIVE_HIGH, IPT_UNKNOWN) PORT_NAME("Memory")
+	PORT_BIT(0x80, IP_ACTIVE_HIGH, IPT_UNKNOWN)
```

Comment:
You shouldn’t use `IPT_UNKNOWN` for things that are known like this.  You should use `IPT_KEYPAD` for data entry keypads that aren’t full keyboards, or `IPT_OTHER` for miscellaneous keys/buttons.

Reply:
Thanks for all your help getting the driver up to scratch, and merged in!